packages/doenetml-worker-rust/src/CoreWorker.ts

```diff
@@ -15,49 +15,204 @@ export interface FlatDastRoot {
+                executeInstructions();
+            }
+        });
+    }
+}
+
+async function executeInstructions() {
+    while (instruction_queue.length > 0) {
+        let next_instruction = instruction_queue.splice(0, 1)[0];
+
+        if (next_instruction.type === "set_source") {
+            await executeSetSource(next_instruction);
+            next_instruction.resolve();
+        } else if (next_instruction.type === "set_flags") {
+            await executeSetFlags(next_instruction);
+            next_instruction.resolve();
+        } else if (next_instruction.type === "return_dast") {
+            let dast = await executeReturnDast(next_instruction);
+            next_instruction.resolve(dast);
+        } else if (next_instruction.type === "terminate") {
+            await executeTerminate(next_instruction);
+            next_instruction.resolve();
         }
+    }
 
-        return JSON.parse(this.doenetCore.return_dast()) as FlatDastRoot;
+    processing_instruction = false;
+}
+
+async function executeSetSource(instruction: SetSourceInstruction) {
+    let worker = instruction.core_worker;
+
+    if (!worker.wasm_initialized) {
+        await init();
+        worker.wasm_initialized = true;
+    }
+
+    if (!worker.doenetCore) {
+        worker.doenetCore = PublicDoenetMLCore.new();
+    }
+
+    worker.doenetCore.set_source(
+        JSON.stringify(instruction.dast),
+        instruction.source,
+    );
+    worker.source_set = true;
+}
+
+async function executeSetFlags(instruction: SetFlagsInstruction) {
+    let worker = instruction.core_worker;
+
+    if (!worker.wasm_initialized) {
+        await init();
+        worker.wasm_initialized = true;
     }
 
-    async terminate() {
-        console.log("CoreWorker.terminate");
-        this.doenetCore?.free();
+    if (!worker.doenetCore) {
+        worker.doenetCore = PublicDoenetMLCore.new();
     }
+
+    worker.doenetCore.set_flags(JSON.stringify(instruction.flags));
+    worker.flags_set = true;
+}
+
+async function executeReturnDast(instruction: ReturnDastInstruction) {
+    let worker = instruction.core_worker;
+    if (!worker.source_set || !worker.flags_set || !worker.doenetCore) {
+        throw Error("Cannot return dast before setting source and flags");
+    }
+
+    try {
+        return JSON.parse(worker.doenetCore.return_dast()) as FlatDastRoot;
+    } catch (err) {
+        console.error(err);
+        throw err;
+    }
+}
+
+async function executeTerminate(instruction: TerminateInstruction) {
+    let worker = instruction.core_worker;
+
+    // TODO: need to call terminate on doenetCore
+    // so that it can attempt to send final state (to allow it to be saved)
+    // before terminating
+
+    worker.doenetCore?.free();
+    worker.doenetCore = undefined;
+
+    // Since we store source and flags in doenetCore,
+    // we lose source and flags when terminating it.
+    worker.source_set = false;
+    worker.flags_set = false;
```

Comment:
This all seems quite excessive...can't you do

```typescript
CoreWorker {
   isProcessingPromise = Promise.resolve();
   async doSomething(){
      const isProcessingPromise = this.isProcesssingPromise;
      let resolve: Function;
      this.isProcessingPromise = new Promise((res) => resolve=res);

      await isProcessingPromise
      // do the stuff ...
      resolve()
   }
}
```

Am I missing something?

Reply:
Interesting. I hadn't thought of that. My understanding of Javascript is that nothing could happen in another call to the worker in between setting `const isProcessingPromise` and setting `this.isProcessingPromise` so there's no race condition here. Is that correct?

Reply:
Yes. The event loop _should_ be uninterruptible...

Reply:
It's complaining that `resolve` is used before it is defined. Any better ideas other than putting a @ts-ignore in front of it?

Reply:
You could try adding a `!` afterwards to assert it's non-null, or you could initialize it with `()=>{}`

Reply:
```typescript
function promiseWithResolver() {
    let resolve: (value: unknown) => void;
    let reject: (reason: unknown) => void;
    const promise = new Promise((res, rej) => {
        resolve = res;
        reject = rej;
    });
    return { promise, resolve: resolve!, reject: reject! };
}
```

Reply:
Handy util. However, when I try this:
```
  const isProcessingPromise = this.isProcessingPromise;

  let { promise, resolve } = promiseWithResolver();
  this.isProcessingPromise = promise;
```
I get this error on the last line:
```
Type 'Promise<unknown>' is not assignable to type 'Promise<void>'.
```
Note: in one case, for `return_dast`, the promise is supposed to return something. Use generics or just skip the utility function?

Reply:
Generics would be fine.
```typescript
function promiseWithResolver<T = unknown>){
  let resolve: (value: T) => void;
  ...
}
```
or something like that :-)

Reply:
Oh, you might also need to type `const promise: Promise<T> = new Promise(...)`

Reply:
Got my promises mixed up with how I was using it for the other way. This promise never returns anything, so I just changed the utility function to
```
function promiseWithResolver() {
    let resolve: (value: void) => void;
    let reject: (reason: unknown) => void;
    const promise: Promise<void> = new Promise((res, rej) => {
        resolve = res;
        reject = rej;
    });
    return { promise, resolve: resolve!, reject: reject! };
}
```